### PR TITLE
feat: support for partially enabled incremental rebuilds

### PR DESCRIPTION
**Pull request by jerrykingxyz**
### <samp>🤖 Generated by Copilot at c49a3fd</samp>

This pull request adds support for incremental rebuild experiments in the `rspack` project. It introduces a new `IncrementalRebuild` struct in Rust and a corresponding `rawIncrementalRebuild` object in TypeScript, with two options: `make` and `emit_asset`. It also updates the existing code in both languages to handle the new shape of the experiment and to check the options before performing certain actions.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at c49a3fd</samp>

*  Introduce a new `IncrementalRebuild` struct to represent the options for incremental rebuild experiments ([link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-c88be6064f9723a97e920eb73113c11ea487b1be35aa57be4889f5caf4fd2a74L2-R10))
*  Replace the `incremental_rebuild` field of the `Experiments` and `RawExperiments` structs with an instance of `IncrementalRebuild` and `RawIncrementalRebuild` respectively ([link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-c88be6064f9723a97e920eb73113c11ea487b1be35aa57be4889f5caf4fd2a74L2-R10), [link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-ee356163b888e01f5855248826002b944820b3eac1c6cb3e5868d2a34c3fc11dL8-R18))
*  Convert a `RawIncrementalRebuild` into an `IncrementalRebuild` by mapping the `make` and `emit_asset` fields ([link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-ee356163b888e01f5855248826002b944820b3eac1c6cb3e5868d2a34c3fc11dL20-R31))
*  Rename the `is_incremental_rebuild` method of the `CompilerOptions` struct to `is_make_use_incremental_rebuild`, to reflect that it only checks the `make` option of the `incremental_rebuild` experiment ([link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-d9aed4cc1e003d1128224454ff63c335e0b0b13df67c908f322af4f39e9e76f7L29-R29), [link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-b6c971a9adb8c246d789a436e09e6bb0f0077c819d437857845e60263aedf103L29-R30))
*  Update the call sites of the renamed method in the `code_splitting_cache.rs`, `hmr.rs`, and `optimizer.rs` files ([link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-d9aed4cc1e003d1128224454ff63c335e0b0b13df67c908f322af4f39e9e76f7L29-R29), [link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-269c100a13b91351b9f928955094fdd7641e93f1f3062c0de339fd041fb4e763L141-R141), [link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-9c68e9018eeda590fa3e631aa7744787eb38b07c08056d5216c57e32c4c9198eL62-R62))
*  Add a condition to check the `emit_asset` option of the `incremental_rebuild` experiment before updating the `emitted_asset_versions` field of the `Compilation` struct in the `compiler/mod.rs` file ([link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-3fbb9f7dfbadceab7b0c89038c54de9851f25ae957a1f91aea05b0e46da2b209L283-R283))
*  Adapt the TypeScript code in the `adapter.ts` file to handle the new shape of the `incremental_rebuild` experiment, which can be either a boolean or an object with `make` and `emit_asset` fields ([link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-5f5bb1aeb2a977ff30c5ab822c79f09b76476c503462939c69beb9e4d7552123L593-R610))
*  Update the JSON schema, TypeScript interface, and Zod schema for the `incremental_rebuild` experiment in the `schema.js`, `types.ts`, and `experiments.ts` files respectively, to allow either a boolean or an object with `make` and `emit_asset` properties ([link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-d7e2c50a4845569f35f387513dc5a6e542e9d5dde107294fd55c5ff4e8a72c88L363-R380), [link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-790d97d1c7c2bfbbe7720a38355ec1e0d462206a5ddb9826c6aa8bff7090ce0fL591-R596), [link](https://github.com/web-infra-dev/rspack/pull/3472/files?diff=unified&w=0#diff-d0d329e7eba5c2bc8a9e487bf452b933834266020393c24d092acd67b8293ce9L6-R14))

</details>

### CODE REVIEW

File: packages/rspack/src/config/zod/experiments.ts [hunk 1]
```diff
@@ -3,7 +3,15 @@ import { z } from "zod";
 export function experiments() {
 	return z.object({
 		asyncWebAssembly: z.boolean().optional(),
-		incrementalRebuild: z.boolean().optional(),
+		incrementalRebuild: z
+			.boolean()
+			.or(
+				z.object({
```
Comment (hyf0) [inline packages/rspack/src/config/zod/experiments.ts:9]: Could you make 
```suggestion
				z.strictObject({
```

**Comment by jerrykingxyz** [comment]
The test of incremental rebuild will be added after the e2e test is completed

File: crates/node_binding/binding.d.ts [hunk 2]
```diff
@@ -504,7 +504,7 @@ export interface RawEntryItem {
 
 export interface RawExperiments {
   lazyCompilation: boolean
-  incrementalRebuild: boolean
+  incrementalRebuild: RawIncrementalRebuild
```
Comment (h-a-n-a) [inline crates/node_binding/binding.d.ts:507]: It seems to be a breaking change, could you make this field a union? You can leverage `napi::Either`

Comment (h-a-n-a) [reply crates/node_binding/binding.d.ts:507]: Didn't see you have changed it on the JS side. Just ignore me

**Review by hyf0 — APPROVED** [review]

**Review by ahabhgk — APPROVED** [review]

**Pull request merged**